### PR TITLE
Added support for RedisURL according to IANA spec

### PR DESCRIPTION
**Pull request by brzuchal**
Implements #55

### CODE REVIEW

File: lib/Connection.php [hunk 1]
```diff
@@ -67,14 +79,52 @@ public function __construct(string $uri)
...
+                    "tcp://%s:%d",
+                    $uri->getHost() ?? self::DEFAULT_HOST,
+                    $uri->getPort() ?? self::DEFAULT_PORT
+                );
+                if ($uri->getPath() !== "/") {
+                    $this->database = (int) \ltrim($uri->getPath(), "/");
+                } else {
+                    $this->database = (int) ($uri->getQueryParameter("db") ?? 0);
+                }
+                if (!empty($uri->getPass())) {
```
Comment (kelunik) [inline lib/Connection.php]: This won't allow the password `0`.

**Comment by brzuchal** [comment]
@kelunik any chance in pushing this?

**Comment by kelunik** [comment]
I plan to merge this, yes. I just forgot about it. I really need a list of open, ready PRs.

File: lib/Connection.php [hunk 2]
```diff
@@ -28,6 +30,12 @@ class Connection
     /** @var string */
     private $uri;
 
+    /** @var string */
+    private $password = '';
```
Comment (kelunik) [inline lib/Connection.php]: Password should default to `null` IMO, does Redis allow empty passwords?

File: lib/Connection.php [hunk 3]
```diff
@@ -67,14 +79,52 @@ public function __construct(string $uri)
         });
     }
 
+    /**
+     * When using the "redis" schemes the URI is parsed according
+     * to the rules defined by the provisional registration documents approved
+     * by IANA. If the URI has a password in its "user-information" part or a
+     * database number in the "path" part these values override the values of
+     * "password" and "database" if they are present in the "query" part.
```
Comment (kelunik) [inline lib/Connection.php]: These are not well-defined. I think we should error if `password` or `database` is present in path and query.

File: lib/Connection.php [hunk 4]
```diff
@@ -224,6 +274,21 @@ public function getState(): int
...
     }
 
+    public function getPassword(): string
+    {
+        return $this->password;
+    }
+
+    public function hasPassword(): bool
+    {
+        return (bool) \strlen($this->database);
```
Comment (kelunik) [inline lib/Connection.php]: Typo, should be `$this->password`

File: test/SelectTest.php [hunk 5]
```diff
@@ -38,4 +49,20 @@ public function select()
...
+    /**
+     * @test
+     */
+    public function selectWithRedisUrl()
+    {
+        Loop::run(function () {
+            $redis1 = new Client("redis://:secret@127.0.0.1:25325/1");
+            $payload = "bar";
+            yield $redis1->set("foobar", $payload);
+            $this->assertEquals($payload, (yield $redis1->get("foobar")));
```
Comment (kelunik) [inline test/SelectTest.php]: There's no need for `()` around `yield` anymore as of PHP 7.

File: lib/Client.php [hunk 6]
```diff
@@ -56,33 +49,27 @@ public function __construct(string $uri)
...
 
-        if ($this->database !== 0) {
+        if ($this->connection->getDatabase() !== 0) {
             $this->connection->addEventHandler("connect", function () {
                 // SELECT must be called for every new connection if another database than 0 is used
                 \array_unshift($this->deferreds, new Deferred);
+                $database = $this->connection->getDatabase();
 
-                return "*2\r\n$6\r\rSELECT\r\n$" . \strlen($this->database) . "\r\n{$this->database}\r\n";
+                return "*2\r\n$6\r\rSELECT\r\n$" . \strlen($database) . "\r\n{$database}\r\n";
```
[the remark this answers is not in the document]
Comment (brzuchal) [reply lib/Client.php]: Agree, already made that change on a different branch.

File: lib/Connection.php (hunk 2 above)
Comment (brzuchal) [reply lib/Connection.php]: [Documentation](https://redis.io/commands/auth) says nothing about an empty password, let's assume it's not allowed.

File: lib/ConnectionConfig.php [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+    /**
+     * @param string $database
+     * @throws ConnectionConfigException
+     */
+    private function setDatabase(string $database)
+    {
+        if (!\is_numeric($database) || (int) $database < 0) {
+            throw new ConnectionConfigException(
+                "Database name should be integer 0-9"
```
Comment (kelunik) [inline lib/ConnectionConfig.php]: There's no upper limit check, but the message says 0-9, could you fix that?

File: lib/ConnectionConfig.php [hunk 8]
```diff
@@ -0,0 +1,203 @@
+<?php declare(strict_types=1);
```
Comment (kelunik) [inline lib/ConnectionConfig.php]: Please remove `declare(strict_types=1);`.

File: lib/ConnectionConfigException.php [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<?php declare(strict_types=1);
```
Comment (kelunik) [inline lib/ConnectionConfigException.php]: Please remove `declare(strict_types=1);`.

**Comment by kelunik** [comment]
I've manually applied parts of this PR, thanks! I hope I didn't miss anything.

**Pull request closed without merging**